x/protorev/keeper/hooks.go

```diff
@@ -0,0 +1,211 @@
+	coins, err := k.poolmanagerKeeper.GetTotalPoolLiquidity(ctx, poolId)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	return coins[0].Amount.Mul(coins[1].Amount), nil
+}
+
+// StoreJoinExitPoolSwaps stores the swaps associated with GAMM join/exit pool messages in the store, depending on if it is a join or exit.
+func (k Keeper) StoreJoinExitPoolSwaps(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, denom string, isJoin bool) {
+	pool, err := k.gammKeeper.GetPoolAndPoke(ctx, poolId)
+	if err != nil {
+		return
+	}
+
+	// Get all the pool coins and iterate to get the denoms that make up the swap
+	coins := pool.GetTotalPoolLiquidity(ctx)
+
+	// Create swaps to backrun being the join coin swapped against all other pool coins
+	for _, coin := range coins {
+		if coin.Denom == denom {
+			continue
+		}
+		// Join messages swap in the denom, exit messages swap out the denom
+		if isJoin {
+			k.StoreSwap(ctx, poolId, denom, coin.Denom)
+		} else {
+			k.StoreSwap(ctx, poolId, coin.Denom, denom)
+		}
+	}
+}
+
+// AfterPoolCreatedWithCoins checks if the new pool should be stored as the highest liquidity pool
+// for any of the base denoms, and stores it if so.
+func (k Keeper) AfterPoolCreatedWithCoins(ctx sdk.Context, poolId uint64) {
+	baseDenoms, err := k.GetAllBaseDenoms(ctx)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting base denoms in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	baseDenomMap := make(map[string]bool)
+	for _, baseDenom := range baseDenoms {
+		baseDenomMap[baseDenom.Denom] = true
+	}
+
+	pool, err := k.poolmanagerKeeper.GetPool(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool in AfterCFMMPoolCreated hook", err)
+		return
+	}
+
+	denoms, err := k.poolmanagerKeeper.RouteGetPoolDenoms(ctx, poolId)
+	if err != nil {
+		ctx.Logger().Error("Protorev error getting pool liquidity in afterPoolCreated", err)
+		return
+	}
+
+	// Pool must be active and the number of denoms must be 2
+	if pool.IsActive(ctx) && len(denoms) == 2 {
+		if _, ok := baseDenomMap[denoms[0]]; ok {
+			k.CompareAndStorePool(ctx, poolId, denoms[0], denoms[1])
+		}
+		if _, ok := baseDenomMap[denoms[1]]; ok {
+			k.CompareAndStorePool(ctx, poolId, denoms[1], denoms[0])
+		}
+	}
+}
+
+// CompareAndStorePool compares the liquidity of the new pool with the liquidity of the stored pool, and stores the new pool if it has more liquidity.
+func (k Keeper) CompareAndStorePool(ctx sdk.Context, poolId uint64, baseDenom, otherDenom string) {
+	storedPoolId, err := k.GetPoolForDenomPair(ctx, baseDenom, otherDenom)
+	if err != nil {
+		// Error means no pool exists for this pair, so we set it
+		k.SetPoolForDenomPair(ctx, baseDenom, otherDenom, poolId)
+		return
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			ctx.Logger().Error("Protorev error recovering from panic in AfterCFMMPoolCreated hook, likely an overflow error", r)
+		}
+	}()
```

Comment:
Could you please elaborate on when and where can this panic? Not immediately clear to me

Reply:
Would be great to have a test covering this as well to show how this can happen

Reply:
Added comment specifying it catches an overflow panic and moved the panic catching into GetComparablePoolLiquidity (where it should belong since moving the multiplication into that function) : https://github.com/osmosis-labs/osmosis/pull/5045/commits/22d528e23f5ff3733c0d7b4c83f03e6914ecf5c4

Reply:
Added test here: https://github.com/osmosis-labs/osmosis/pull/5045/commits/ed66bce110964f30b32d9d2a2c88922c522d5556

Reply:
The simulator sometimes fails due to incorrect setup or false positives. Usually, we inspect the logs, and if it is something unrelated, simply rerunning it helps.